test/e2e/all_in_one_test.go

```diff
@@ -935,3 +1155,48 @@ func getPreviousGitTag(path string, cur semver.Version) (semver.Version, error)
 	}
 	return tags[curIndex-1], nil
 }
+
+// getKongProxyIP takes a Service with Kong proxy ports and returns and its IP, or fails the test if it cannot
+func getKongProxyIP(ctx context.Context, t *testing.T, env environments.Environment, svc *corev1.Service) string {
+	proxyIP := ""
+	require.NotEqual(t, svc.Spec.Type, svc.Spec.ClusterIP)
+	if svc.Spec.Type == corev1.ServiceTypeLoadBalancer {
+		if len(svc.Status.LoadBalancer.Ingress) > 0 {
+			proxyIP = svc.Status.LoadBalancer.Ingress[0].IP
+		}
+	}
+	// the above failed to find an address. either the LB didn't provision or we're using a NodePort
+	if proxyIP == "" {
+		var port int32
+		for _, sport := range svc.Spec.Ports {
+			if sport.Name == "kong-proxy" || sport.Name == "proxy" {
+				port = sport.NodePort
+			}
+		}
+		var extAddrs []string
+		var intAddrs []string
+		nodes, err := env.Cluster().Client().CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+		require.NoError(t, err)
+		for _, node := range nodes.Items {
+			for _, naddr := range node.Status.Addresses {
+				if naddr.Type == corev1.NodeExternalIP {
+					extAddrs = append(extAddrs, naddr.Address)
+				}
+				if naddr.Type == corev1.NodeInternalIP {
+					extAddrs = append(intAddrs, naddr.Address)
+				}
+			}
+		}
+		// local clusters (KIND, minikube) typically provide no external addresses, but their internal addresses are
+		// routeable from their host. We prefer external addresses if they're available, but fall back to internal
+		// in their absence
+		if len(extAddrs) > 0 {
+			proxyIP = fmt.Sprintf("%v:%v", extAddrs[0], port)
+		} else if len(intAddrs) > 0 {
+			proxyIP = fmt.Sprintf("%v:%v", intAddrs[0], port)
+		} else {
+			assert.Fail(t, "both extAddrs and intAddrs are empty")
+		}
+	}
+	return proxyIP
+}
```

Comment:
It seems like we use a case statement here to differentiate the logic paths, instead of using initialization logic followed by state checks to do so. This could help make the code a bit easier to read and allow us to refactor the longer nodeport logic into it's own clear function.

Suggestion: refactor into case statement with sub-functions

Reply:
Maybe--this wasn't really in scope here, it didn't get added, just extracted out of `verifyIngress()`

Reply:
Ok, I dont' consider this a blocker.